providers/vultr/vultr.go

```diff
@@ -0,0 +1,342 @@
+			},
+		})
+	}
+
+	for _, mod := range create {
+		r := toVultrRecord(dc, mod.Desired)
+		corrections = append(corrections, &models.Correction{
+			Msg: mod.String(),
+			F: func() error {
+				return api.client.CreateDNSRecord(dc.Name, r.Name, r.Type, r.Data, r.Priority, r.TTL)
+			},
+		})
+	}
+
+	for _, mod := range modify {
+		id := mod.Existing.Original.(*vultr.DNSRecord).RecordID
+		r := toVultrRecord(dc, mod.Desired)
+		r.RecordID = id
+		corrections = append(corrections, &models.Correction{
+			Msg: fmt.Sprintf("%s; Vultr RecordID: %v", mod.String(), id),
+			F: func() error {
+				return api.client.UpdateDNSRecord(dc.Name, *r)
+			},
+		})
+	}
+
+	return corrections, nil
+}
+
+// GetNameservers gets the Vultr nameservers for a domain
+func (api *VultrApi) GetNameservers(domain string) ([]*models.Nameserver, error) {
+	return models.StringsToNameservers(defaultNS), nil
+}
+
+// EnsureDomainExists adds a domain to the Vutr DNS service if it does not exist
+func (api *VultrApi) EnsureDomainExists(domain string) error {
+	ok, err := api.isDomainInAccount(domain)
+	if err != nil {
+		return err
+	}
+
+	if !ok {
+		// Vultr requires an initial IP, use a dummy one
+		err := api.client.CreateDNSDomain(domain, "127.0.0.1")
+		if err != nil {
+			return err
+		}
+
+		ok, err := api.isDomainInAccount(domain)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			return fmt.Errorf("Unexpected error adding domain %s to Vultr account", domain)
+		}
+	}
+
+	return nil
+}
+
+func (api *VultrApi) isDomainInAccount(domain string) (bool, error) {
+	domains, err := api.client.GetDNSDomains()
+	if err != nil {
+		return false, err
+	}
+
+	var vd *vultr.DNSDomain
+	for _, d := range domains {
+		if d.Domain == domain {
+			vd = &d
+		}
+	}
+
+	if vd == nil {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// toRecordConfig converts a Vultr DNSRecord to a RecordConfig
+func toRecordConfig(dc *models.DomainConfig, r *vultr.DNSRecord) (*models.RecordConfig, error) {
+	// Turns r.Name into a FQDN
+	// Vultr uses "" as the apex domain, instead of "@", and this handles it fine.
+	name := dnsutil.AddOrigin(r.Name, dc.Name)
+
+	data := r.Data
+	// Make target into a FQDN if it is a CNAME, NS, MX, or SRV #rtype_variations
+	if r.Type == "CNAME" || r.Type == "NS" || r.Type == "MX" {
+		if !strings.HasSuffix(data, ".") {
+			data = data + "."
+		}
+		data = dnsutil.AddOrigin(data, dc.Name)
+	}
+	// Remove quotes if it is a TXT #rtype_variations
+	if r.Type == "TXT" {
+		if !strings.HasPrefix(data, `"`) || !strings.HasSuffix(data, `"`) {
+			return nil, errors.New("Unexpected lack of quotes in TXT record from Vultr")
+		}
+		data = data[1 : len(data)-1]
+	}
+
+	rc := &models.RecordConfig{
+		NameFQDN: name,
+		Type:     r.Type,
+		Target:   data,
+		TTL:      uint32(r.TTL),
+		Original: r,
+	}
+
+	// #rtype_variations
+	if r.Type == "MX" {
+		rc.MxPreference = uint16(r.Priority)
+	}
+
+	// #rtype_variations
+	if r.Type == "SRV" {
```

Comment:
I believe there are some helper functions in the models package for splitting out srv/caa records from a single field representation. Is vultr doing something odd so that those don't work?

Reply:
I'll have to look into that, but Vultr does quite a few funny things, so it may be better to manually handle it.

Reply:
The helper functions do not use the right order for Vultr.